src/trace.rs

```diff
@@ -0,0 +1,331 @@
+        self.run(max_steps);
+        let lps = self.root.accumulate(|n| {
+            let n_r = &n.0.read().expect("poisoned TraceNode");
+            Term::alpha(term, &n_r.term).map(|_| n_r.log_p)
+        });
+        if lps.is_empty() {
+            f64::NEG_INFINITY
+        } else {
+            logsumexp(&lps)
+        }
+    }
+}
+
+/// A single node in a Trace.
+#[derive(Debug, Clone)]
+struct TraceNodeStore {
+    term: Term,
+    state: TraceState,
+    log_p: f64,
+    depth: usize,
+    parent: Option<Weak<RwLock<TraceNodeStore>>>,
+    children: Vec<TraceNode>,
+}
+
+/// The state of a Node in a Trace
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum TraceState {
+    Start,
+    Normal,
+    Unobserved,
+    TooBig,
+}
+
+#[derive(Debug, Clone)]
+pub struct TraceNode(Arc<RwLock<TraceNodeStore>>);
+impl TraceNode {
+    fn new(
+        term: Term,
+        state: TraceState,
+        log_p: f64,
+        depth: usize,
+        parent: Option<&TraceNode>,
+        children: Vec<TraceNode>,
+    ) -> TraceNode {
+        let parent = parent.map(|n| Arc::downgrade(&n.0));
+        TraceNode(Arc::new(RwLock::new(TraceNodeStore {
+            term,
+            state,
+            log_p,
+            depth,
+            parent,
+            children,
+        })))
+    }
+    pub fn state(&self) -> TraceState {
+        self.0.read().expect("poisoned TraceNode").state
+    }
+    pub fn term(&self) -> Term {
+        self.0.read().expect("poisoned TraceNode").term.clone()
+    }
+    pub fn log_p(&self) -> f64 {
+        self.0.read().expect("poisoned TraceNode").log_p
+    }
+    pub fn depth(&self) -> usize {
+        self.0.read().expect("poisoned TraceNode").depth
+    }
+    pub fn parent(&self) -> Option<TraceNode> {
+        self.0
+            .read()
+            .expect("poisoned TraceNode")
+            .parent
+            .as_ref()
+            .and_then(|w| w.upgrade())
+            .map(TraceNode)
+    }
+    pub fn children(&self) -> Vec<TraceNode> {
+        self.0.read().expect("poisoned TraceNode").children.clone()
+    }
+    pub fn is_leaf(&self) -> bool {
+        self.0
+            .read()
+            .expect("poisoned TraceNode")
+            .children
+            .is_empty()
+    }
+
+    fn accumulate<T, F>(&self, condition: F) -> Vec<T>
+    where
+        F: Fn(&TraceNode) -> Option<T>,
+    {
+        let mut values = Vec::new();
+        self.walk(|n| {
+            if let Some(v) = condition(n) {
+                values.push(v)
+            }
+        });
+        values
+    }
+    fn walk<F>(&self, mut callback: F)
+    where
+        F: FnMut(&TraceNode),
+    {
+        self.walk_helper(&mut callback)
+    }
+    fn walk_helper<F>(&self, callback: &mut F)
+    where
+        F: FnMut(&TraceNode),
+    {
+        callback(self);
+        for child in &self.0.read().expect("poisoned TraceNode").children {
+            child.walk_helper(callback)
+        }
+    }
```

Comment:
`accumulate`, `walk`, and `walk_helper` feel like they should be part of some sort of trait implementation, like `Traversable` in Haskell. Might this just be `Iterator`?

Reply:
I'll add an `Iterator` implementation, but I think it's worth keeping `accumulate` and `walk` here — they are faster because no cloning is necessary during iteration (because of how the lifetimes and the read guards work, an `Iterator` over `TraceNode`s would mandate cloning) and the stack can be used directly rather than having to store DFS/BFS state in an `Iter`-like struct (unlike with a `Vec` and other linear data structures, an iterator here would store `O(n)` space, not `O(1)`).